### DISCUSSION TITLE
Build Image tar gz files

### DISCUSSION THREAD

**Discussion by lantos1618** [moby/moby#44048] [General]
Hi was the intention of [imageBuild](https://docs.docker.com/engine/api/v1.41/#tag/Image/operation/ImageBuild) to be so that a user can build dockerfiles with dependancies? 

Currently a user can only send a Dockerfile that has been archived and compressed into a dockerfile.tar.gz 
but if the Dockerfile has external dependencies it would be good to allow users to archive and compress them alongside the Dockerfile and have them in there.

BuildFolder
 - Dockerfile
 - somedependancy.file

Can a user compress the build folder and then have an image built from that?

**Comment by thaJeztah** [comment]
Yes, you should be able to send the whole buildcontext (all files used, together with the Dockerfile in it). Effectively, that's what the classic builder does when you run `docker build .`; it creates a tar from the build context (`.` - current directory in that example), and uploads it to the daemon; the daemon then extracts it, and starts a build.

When Building with BuildKit, files are sent using an interactive session (the daemon requests files from the client that the Dockerfile uses, which then get sent over if needed)

**Reply by lantos1618** [reply]
Hey thanks for the reply. You're right. I thought I was doing everything right but found out that the tar.gz library I'm using isn't functioning the same as "github.com/docker/docker/pkg/archive" So the tar.gz I'm using appears to be missing something. 


example repo.
https://github.com/lantos-lgtm/docker-tar-gz-issue-test
Is "github.com/docker/docker/pkg/archive"  doing anything more special then just tar + gunziping?

**Reply by lantos1618** [reply]
> Hey thanks for the reply. You're right. I thought I was doing everything right but found out that the tar.gz library I'm using isn't functioning the same as "github.com/docker/docker/pkg/archive" So the tar.gz I'm using appears to be missing something.
> 
> example repo. https://github.com/lantos-lgtm/docker-tar-gz-issue-test Is ``"github.com/docker/docker/pkg/archive"`` doing anything more special then just tar + gunziping?

Okay so I found the reason.

When a user tries to compress a folder manually or by using a package other then ``"github.com/docker/docker/pkg/archive"``
the initial directory is included.

example of what is going on
example folder structure
- testDocker
  - Dockerfile
  - index.html

running  `` tar -czvf testDocker.cmd.tar.gz testDocker`` or using a package other then ``"github.com/docker/docker/pkg/archive"``
the defaults will produce inside the tar file like so
``testDocker.cmd.tar.gz``
- ./testDocker/Dockerfile
- ./testDocker/index.html

However the docker api ```imageBuild()``` expects the Docker entry point to be on the first level like so
``testDocker.go.tar.gz``
- Dockerfile
- index.html

this can be achieved by stripping the first folder with `` tar -czvf testDocker.cmd.stripped.tar.gz -C testDocker`` like so
``testDocker.cmd.stripped.tar.gz ``
- ./Dockerfile
- ./index.html


Would it be a good feature request to add the following logic to the imageBuild?
psuedo code
```
if (tarFile does not have "./Dockerfile" and has one folder) {
 change build directory into that folder
}
continue build process
```
